### PR TITLE
modify the framework of OSPI Flash driver

### PR DESCRIPTION
**Pull request by rogeryou**
The description of this commit is [Suggestion on the framework of Flash driver #46399](https://github.com/zephyrproject-rtos/zephyr/issues/46399)

### CODE REVIEW

**Comment by GeorgeCGV** [comment]
It is a great idea to refactor NOR-Flash support. It is not really a refactoring, it looks more like a complete rework.

I like the split of the NOR-Flash driver from the OSPI bus. The same shall be done for QSPI one day. Eventually, QSPI and OSPI shall share the NOR-Flash driver implementation, as there is major functionality crossover. Considering that Zephyr tries to be similar to Linux Kernel, we shall look more into the existing Linux Kernel SPI NOR-Flash implementation.

However, I don't think that there is a need for defines to be duplicated by `ospi_nor.h` instead of extending the existing `spi_nor.h`. Opcodes are more or less related to the shared JEDEC Standard (JESD216).

Unfortunately, the PR removes the current ability of OSPI to work with quad and dual flashes in `1-1-2`, `1-2-2`, `1-1-4`, `1-4-4` modes with 3-byte addressing support due to hardcoded opcodes usage of 4-byte addressing. 

Overall, this PR negates existing efforts. 

My personal opinion is that the PR shall include the functionality of the existing driver as well:

- Work with 3-byte and 4-byte addressing.
- Allow detection of read opcode and dummy cycles from SFDP: BFP. Alternative, allow to set it in DTS (same way as with writeoc).
- Support `quad-enable-requirements` in DTS, with bypass over SFDP: BFP DW15.
- Support enabling of a quad mode via different `quad-enable-requirements` values.

**Comment by rogeryou** [comment]
> 

Thanks for your review.

Our initial idea was to build an independent driver implementation for spi/qspi/ospi flash, such as spi_nor.c for SPI NOR Flash, qspi_nor.c for QSPI NOR Flash, ospi_nor.c for OSPI NOR Flash.

We suggest that there should be only these three driver files about nor flash in the flash folder, and others, such as flash_ stm32_ ospi. c, nrf_ qspi_ nor. c and other underlying drivers related to qspi/ospi controllers are placed in the SPI folder.

So in ospi_ nor.c, we removed the codes that have nothing to do with OSPI NOR Flash, such as quad and 3 byte /4 byte addresses.

But you suggest that QSPI and OSPI shall share the NOR-Flash driver implementation，this is also a reasonable solution， 
If necessary, we can modify the code as you suggest.

File: boards/arm/stm32l562e_dk/stm32l562e_dk_common.dtsi [hunk 1]
```diff
@@ -139,13 +139,16 @@
 	status = "okay";
 
 	mx25lm51245: ospi-nor-flash@0 {
-		compatible = "st,stm32-ospi-nor";
+		compatible = "jedec,ospi-nor";
 		label = "MX25LM51245";
 		reg = <0>;
-		ospi-max-frequency = <DT_FREQ_M(50)>;
+		spi-max-frequency = <DT_FREQ_M(50)>;
```
Comment (manoj153) [inline boards/arm/stm32l562e_dk/stm32l562e_dk_common.dtsi:145]: The Example does not match the `properties:` `spi-max-frequency`

On a side note, is not the former`ospi-max-frequency` sounds better?

File: dts/bindings/flash_controller/jedec,ospi-nor.yaml [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+    Representation of a serial flash on a octospi bus:
+
+        mx25lm51245: ospi-nor-flash@0 {
+                compatible = "st,stm32-ospi-nor";
+                label = "MX25LM512";
+                reg = <0>;
+                data-mode = <OSPI_OPI_MODE>; /* access on 8 data lines */
+                data-rate = <OSPI_DTR_TRANSFER>; /* access in DTR */
+                ospi-max-frequency = <DT_FREQ_M(50)>;
```
Comment (manoj153) [inline dts/bindings/flash_controller/jedec,ospi-nor.yaml:15]: Not valid probably missed to update, the property is now renamed to `spi-max-frequency`

File: dts/bindings/flash_controller/jedec,ospi-nor.yaml [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+      description: Maximum clock frequency of device's OSPI interface in Hz
+    duplex:
+      type: int
+      required: true
+      description: Maximum clock frequency of device's OSPI interface in Hz
+    frame-format:
+      type: int
+      required: true
+      description: Maximum clock frequency of device's OSPI interface in Hz
+    label:
```
Comment (manoj153) [inline dts/bindings/flash_controller/jedec,ospi-nor.yaml:41]: Description are wrong

**Pull request closed without merging**